docs/proposals/vip-4.md

```diff
@@ -0,0 +1,248 @@
+|                      `preSchemaRegistration` |        ❌         |         ✅         |   1️⃣    |
+|                     `postSchemaRegistration` |        ✅         |         ✅         |   1️⃣    |
+|                    `preStoreVersionCreation` |        ✅         |         ✅         |    ✅     |
+|                   `postStoreVersionCreation` |        ✅         |         ✅         |    ✅     |
+|  `preStartOfStoreVersionIngestionForDaVinci` |       2️⃣        |        2️⃣        |   2️⃣    |
+| `postStartOfStoreVersionIngestionForDaVinci` |       2️⃣        |        2️⃣        |   2️⃣    |
+|          `postStoreVersionLeaderReplication` |        ✅         |         ❌         |    ❌     |
+|                        `preStoreVersionSwap` |        ✅         |         ✅         |    ✅     |
+|                       `postStoreVersionSwap` |        ✅         |         ✅         |    ✅     |
+|                            `preEndOfPushJob` |        ✅         |         ✅         |    ✅     |
+|                           `postEndOfPushJob` |        ✅         |         ✅         |    ✅     |
+
+The cells with numbers in them represent special cases, which are explained below:
+
+1️⃣ : Schema registration hooks in push jobs could only be invoked in cases where auto-registration is enabled and the 
+new schema originates from the push job itself, whereas schema registrations which are performed directly on the 
+controller could not trigger the hook.
+
+2️⃣ : The hook for the start of ingestion for Da Vinci Clients is tricky for a few reasons. The start of ingestion is 
+controlled by updating the Meta Store, which is a non-replicated system store updated by child controllers, so those
+must be involved (either by running the hook there in the first place, or by having some mechanism that enables the
+parent or VPJ to inform the child controllers of when the system store is eligible for getting updated, such as by
+adding a new field to the `AddVersion` admin channel command). However, see Rollback Support below for why running hooks
+in the child controller may be insufficient.
+
+### Rollback Support
+
+In order to support the ability for the hooks which return the `StoreVersionLifecycleEventOutcome` to rollback, the most
+natural way to achieve this is likely to involve the parent controller, either by having those hooks run there in the 
+first place, or by having a propagation mechanism to it:
+
+* If the hooks are executed in the child controller, the propagation mechanism might be to extend the job status check 
+  which the parent controller does periodically in order for the child to inform the parent of the need to rollback, or 
+  else build a new mechanism for the child to directly interact with the parent (or even with other child controllers 
+  directly...).
+
+* If the hooks are executed in VPJ, then it would need to interact with the parent via the controller client to trigger 
```

Comment:
Orchestrator can be different from hook executor.
BTW, for the delayed promotion to other regions, I assume we are going to halt the push job and still report the final status in the push job by either abort or rolling forward, right?
There are some costs with this approach since VPJ executor will take more resources than before, and if we don't halt push job, the final status won't be easily observed by the end user.
Let us call out the additional cost/impact to the existing components if applicable.

Reply:
> Orchestrator can be different from hook executor.

Can you elaborate on what you're thinking here?

> BTW, for the delayed promotion to other regions, I assume we are going to halt the push job and still report the final status in the push job by either abort or rolling forward, right?
> There are some costs with this approach since VPJ executor will take more resources than before, and if we don't halt push job, the final status won't be easily observed by the end user.
> Let us call out the additional cost/impact to the existing components if applicable.

Hmm... I think we would definitely want the VPJ driver/executor (which we run in Azkaban) to stall as long as the hooks are instructing to wait, so that the final status of the job reflects the reality. But the driver job is pretty lightweight from my understanding, so I don't expect the cost of that running longer to be significant... the costly part, from my understanding, is the (MR) data push job but that would not take longer due to hooks.

Reply:
The delay in the driver will introduce some cost as each driver is an independent process in Azkaban executor, which pre-allocates some fixed amount of memory for JVM heap, and too many hanging processes might kill the Azkaban executors.
I don't have the data points and these are purely theoretical analysis.

Reply:
So the reasoning here that increasing wait time on pushes might cause too many Azkaban jobs to run simultaneously and kill the executors ? If thats the case, then random slowness in vpj or fabric connection issues would cause same thing, which we do not see that often.